### PR TITLE
doc: connectivity: Bluetooth: add LLCP fly-in

### PR DESCRIPTION
**Pull request by erbr-ot**
Adding a brief fly-in of the main concepts of the
implementation of link layer control procedures handling

### CODE REVIEW

**Review by cvinayak — DISMISSED** [review]

**Comment by erbr-ot** [comment]
just cleaned up some weird '?' characters that had copy/paste'd their way into the mix...

**Review by cvinayak — DISMISSED** [review]
Did a quick read through: https://builds.zephyrproject.io/zephyr/pr/78469/docs/connectivity/bluetooth/bluetooth-ctlr-arch.html#link-layer-control-procedures

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 1]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+   * LLCP main state machine entry/tick'er
+   * From ull_peripheral.c/ull_central.c::ticker_cb via ull_conn_llcp()
+
+- rx_demux context to drive ull_cp_tx_ack() and ull_cp_rx()
+   * LLCP tx ack handling and PDU reception
+   * From ull_conn.c::ull_conn_rx()
+   * Handles passing PDUs into running procedures as well as possibly initiating remote procedures
+
+- Emulated versions of rx_demux/prepare context used in unit tests - testing ONLY procedure PDU flow
+   * event_prepare()/event_done() helpers emultating LLL prepare/done flow
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:162]:
```suggestion
   * event_prepare()/event_done() helpers emulating LLL prepare/done flow
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 2]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+
+- struct llcp_mem_pool
+   * Mem pool used to implement procedure contexts resource - instantiated in both a local and a remote version
+   * Used through ull_llcp.c::create_procedure()
+
+- Misc. pdu gymnastics
+   * Encoding and decoding of control pdus done by ull_llcp_pdu.c::llcp_pdu_encode/decode_<PDU>()
+   * Misc. pdu validation handled by ull_llcp.c::pdu_validate_<PDU>() via ull_llcp.c::pdu_is_valid()
+
+LLCP local and remote request/procedure STMs
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:189]:
```suggestion
LLCP local and remote request/procedure state machines
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 3]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+
+- ull_llcp_remote.c
+   * Remote versions of the above
+   * Naming concept: rr_<...> => remote request machine
+   * Also handling of remote procedure initiation by llcp_rx_new()
+   * Misc. procedure collision handling (in rr_st_idle())
+
+- ull_llcp_common/conn_upd/phy/enc/cc/chmu.c
+   * Individual procedure implementations (ull_llcp_common.c collects the simpler ones)
+   * Naming concept: lp => local initiated procedure, rp_<...> => remote initiated procedure
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:206]:
```suggestion
   * Naming concept: lp_<...>  => local initiated procedure, rp_<...> => remote initiated procedure
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 4]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+   * Handling of procedure flow from init (possibly through instant) to completion and host notification if applicable
+
+Miscelaneous concepts
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:209]:
```suggestion
Miscellaneous concepts
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 5]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+   * Spec states rules for how to handle/resolve collisions when they happen
+
+- Termination handling
+   * Specific rules apply re. how termination is handled.
+   * Since we have resource handling re. procedure contexts and terminate must always be available this is handled as a special case
+   * Note also - there are misc. cases where connection termination is triggered on invalid peer behavior
+
+- New remote procedure handling
+   * Table new_proc_lut[] maps LLCP PDUs to procedures/roles used in llcp_rr_new()
+   * Note - there can only ever be ONE remote procedure in the remote procedure queue
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:224]: The one remote procedure in the queue is per connection, right?

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 6]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+
+- Misc. minors
+   * pause/resume concepts - there are two (see spec for details)
+   * procedure execution can be paused be encryption procedure
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:228]:
```suggestion
   * procedure execution can be paused by the encryption procedure
```

**Review by kruithofa — CHANGES_REQUESTED** [review]

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 7]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+- rx_demux context to drive ull_cp_tx_ack() and ull_cp_rx()
+   * LLCP tx ack handling and PDU reception
+   * From ull_conn.c::ull_conn_rx()
+   * Handles passing PDUs into running procedures as well as possibly initiating remote procedures
+
+- Not a normal execution context, but for unit testing: Emulated versions of rx_demux/prepare context used in unit tests - testing ONLY procedure PDU flow
+   * event_prepare()/event_done() helpers emulating LLL prepare/done flow
+   * lt_rx()/lt_tx() 'lower tester' emulation of rx/tx
+   * ut_rx_node() 'upper tester' emulation of notification flow handling
+   * Bunch of helpers to generate and parse PDUs, as well as misc. mocked ull_stuff()
```
Comment (thoh-ot) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:165]: Wouldn't this section be better suited together with the other "unit test concepts"?

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 8]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+===============================
+
+- struct llcp_struct
+   * Main LLCP data store
+   * Defined in ull_conn_types.h and declared as part of struct ll_conn
+   * Holds local and remote procedure request queues as well as conn specific LLCP data
+   * Basic conn-level abstraction
+
+- struct proc_ctx
+   * General procedure context data, contains misc. procedure data and state as well as sys_snode_t for queueing
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:171]:
```suggestion
   * General procedure context data, contains miscellaneous procedure data and state as well as sys_snode_t for queueing
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 9]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+   * Defined in ull_llcp_internal.h, declared/instantiated through ull_llcp.c::create_procedure()
+   * Also holds node references used in tx_ack as well as rx_node retention mechanisms
+
+- struct llcp_mem_pool
+   * Mem pool used to implement procedure contexts resource - instantiated in both a local and a remote version
+   * Used through ull_llcp.c::create_procedure()
+
+- Misc. pdu gymnastics
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:179]:
```suggestion
- Miscellaneous pdu gymnastics
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 10]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+   * Encoding and decoding of control pdus done by ull_llcp_pdu.c::llcp_pdu_encode/decode_<PDU>()
+   * Misc. pdu validation handled by ull_llcp.c::pdu_validate_<PDU>() via ull_llcp.c::pdu_is_valid()
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:181]:
```suggestion
   * Miscellaneous pdu validation handled by ull_llcp.c::pdu_validate_<PDU>() via ull_llcp.c::pdu_is_valid()
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 11]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+   * State machine handling local initiated procedures
+   * Naming concept: lr _<...> => local request machine
+   * Local procedure queue handling
+   * Local run/rx/tx_ack switch
+
+- ull_llcp_remote.c
+   * Remote versions of the above
+   * Naming concept: rr_<...> => remote request machine
+   * Also handling of remote procedure initiation by llcp_rx_new()
+   * Misc. procedure collision handling (in rr_st_idle())
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:196]:
```suggestion
   * Miscellaneous procedure collision handling (in rr_st_idle())
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 12]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+
+- Procedure collision handling
+   * See BT spec for explanation
+   * Basically some procedures can exist in parallel but some can't - for instance only one instant based at a time
+   * Spec states rules for how to handle/resolve collisions when they happen
+
+- Termination handling
+   * Specific rules apply re. how termination is handled.
+   * Since we have resource handling re. procedure contexts and terminate must always be available this is handled as a special case
+   * Note also - there are misc. cases where connection termination is triggered on invalid peer behavior
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:214]:
```suggestion
   * Note also - there are miscellaneous cases where connection termination is triggered on invalid peer behavior
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 13]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+
+- Rx node handling is mocked
+   * Different configs are handled by separate conf files (see ctrl_conn_update for example)
+   * ZTEST(periph_rem_no_param_req, test_conn_update_periph_rem_accept_no_param_req)
+
+- Emulated versions of rx_demux/prepare context used in unit tests - testing ONLY procedure PDU flow
+   * event_prepare()/event_done() helpers emulating LLL prepare/done flow
+   * lt_rx()/lt_tx() 'lower tester' emulation of rx/tx
+   * ut_rx_node() 'upper tester' emulation of notification flow handling
+   * Bunch of helpers to generate and parse PDUs, as well as misc. mocked ull_stuff()
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:245]:
```suggestion
   * Bunch of helpers to generate and parse PDUs, as well as miscellaneous mocked ull_stuff()
```

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 14]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+
+- New remote procedure handling
+   * Table new_proc_lut[] maps LLCP PDUs to procedures/roles used in llcp_rr_new()
+   * Note - for any given connection, there can only ever be ONE remote procedure in the remote procedure queue
+
+- Miscellaneous minors
+   * pause/resume concepts - there are two (see spec for details)
+   * procedure execution can be paused by the encryption procedure
+   * data TX can be paused by PHY, DLE and ENC procedure
+   * RX node retention - ensures no waiting for alloc of RX node when needed for notification
```
Comment (kruithofa) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:224]:
```suggestion
   * RX node retention - ensures no waiting for allocation of RX node when needed for notification
```

**Review by kruithofa — CHANGES_REQUESTED** [review]

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst (hunk 8 above)
Comment (erbr-ot) [reply doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst]: ack

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst (hunk 9 above)
Comment (erbr-ot) [reply doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst]: ack

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst (hunk 11 above)
Comment (erbr-ot) [reply doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst]: ack

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst (hunk 12 above)
Comment (erbr-ot) [reply doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst]: ack

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst (hunk 14 above)
Comment (erbr-ot) [reply doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst]: ack

**Comment by erbr-ot** [comment]
I have ack.'ed misc. suggestions ;-)
Thanks Andries

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst (hunk 13 above)
Comment (erbr-ot) [reply doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst]: ack

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst [hunk 15]
```diff
@@ -135,6 +135,117 @@ Execution Priorities
...
+
+Miscellaneous unit test concepts
+================================
+
+- Individual ZTEST unit test for each procedure
+   * zephyr/tests/bluetooth/controller/ctrl_<proc>
+
+- Only PDU flow is tested (as a main rule)
+   * Helpers emulate prepare/tx_ack/rx by event_prepare(), event_tx_ack() and event_done()
+   * PDU flow to/from controller is emulated by lt_rx/lt_tx and ut_rx interfacing with test rx/tx queues
```
Comment (thoh-ot) [inline doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst:235]: Isn't this just a "compacted" version of the details in L241-L245?

**Review by kruithofa — DISMISSED** [review]

File: doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst (hunk 15 above)
Comment (erbr-ot) [reply doc/connectivity/bluetooth/bluetooth-ctlr-arch.rst]: Agree - removed 233-235

**Review by kruithofa — APPROVED** [review]

**Review by thoh-ot — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]

**Pull request merged**